In [1]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [2]:
# Refs: Module 20/Day1/Act-01
# Read the CSV file from the Resources folder into a Pandas DataFrame
file_path = Path("Resources/lending_data.csv")
df_lending = pd.read_csv(file_path)

# Review the DataFrame
df_lending.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [3]:
# Get the columns names
df_lending.columns

Index(['loan_size', 'interest_rate', 'borrower_income', 'debt_to_income',
       'num_of_accounts', 'derogatory_marks', 'total_debt', 'loan_status'],
      dtype='object')

In [4]:
# Refs: Module 20/Day1/Act-01
# Separate the data into labels and features

# Separate the y variable, the labels
y = df_lending["loan_status"]

# Separate the X variable, the features
X = df_lending[['loan_size', 'interest_rate', 'borrower_income', 'debt_to_income',
       'num_of_accounts', 'derogatory_marks', 'total_debt']]

In [5]:
# Refs: Module 20/Day1/Act-01
# Review the y variable Series
y[:5]

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [6]:
# Review the X variable DataFrame. Select 10 random samples from data stored in variable "X"
X.sample(10)

loan_size  interest_rate  borrower_income  debt_to_income  \
53398     8400.0          6.691            43600        0.311927   
75444    18100.0         10.836            82600        0.636804   
16386    10300.0          7.486            51000        0.411765   
40677     9400.0          7.105            47500        0.368421   
28657     9900.0          7.330            49600        0.395161   
1202     10400.0          7.527            51400        0.416342   
62903     6700.0          5.968            36800        0.184783   
7610      9500.0          7.180            48200        0.377593   
21183     9000.0          6.968            46200        0.350649   
68826     8500.0          6.730            43900        0.316629   

       num_of_accounts  derogatory_marks  total_debt  
53398                3                 0       13600  
75444               11                 2       52600  
16386                4                 1       21000  
40677                3                 0       17500  
28657                4                 0       19600  
1202                 4                 1       21400  
62903                1                 0        6800  
7610                 4                 0       18200  
21183                3                 0       16200  
68826                3                 0       13900

### Step 3: Split the data into training and testing datasets by using `train_test_split`.

In [7]:
# Refs: Module 20/Day1/Act-04
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, 
                                                    y, 
                                                    random_state=1, 
                                                    stratify=y)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [8]:
# Refs: Module 20/Day1/Act-02 & 04
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
model = LogisticRegression(solver='lbfgs', max_iter=200, random_state=1)

# Fit the model using training data
model.fit(X_train, y_train)

LogisticRegression(max_iter=200, random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [9]:
# Refs: Module 20/Day1/Act-02
# Make a prediction using the testing data

# Make a prediction using the X_test
y_predicted = model.predict(X_test)

# Create a DataFrame for the predicted y values
df_y_predicted = pd.DataFrame({"y_test": y_test, "predicted": y_predicted})
df_y_predicted.sample(10)

y_test  predicted
55266       0          0
19977       0          0
11307       0          0
59859       0          0
37851       0          0
13248       0          0
28672       0          0
28375       0          0
14734       0          0
13807       0          0

### Step 3: Evaluate the model’s performance by doing the following:

* Generate a confusion matrix.

* Print the classification report.

In [10]:
# Refs: Module 20/Day1/Act-05
# Generate a confusion matrix for the model
confusion_matrix_model = confusion_matrix(y_test, y_predicted)
print("Confusion Matrix:")
print(confusion_matrix_model)

Confusion Matrix:
[[18679    80]
 [   67   558]]


In [11]:
# Refs: Module 20/Day1/Act-05
# Print the classification report for the model
print("The classification report for the model:")
print(classification_report(y_test, y_predicted))

The classification report for the model:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     18759
           1       0.87      0.89      0.88       625

    accuracy                           0.99     19384
   macro avg       0.94      0.94      0.94     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** When compared to the original dataset, there are more healthy loans than unhealthy loans. The model demonstrates a high accuracy of 99%. The precision score for '0' (healthy loans) is 100%, while the precision for '1' (unhealthy loans) is 87%. Additionally, the recall score is 100% for predicting '0' labels and 89% for high-risk loans labeled '1'.

---